### PR TITLE
Introduction to Progressive Web Apps (Push Notifications) - Part 3

### PR DESCRIPTION
**Pull request by unicodeveloper**

### CODE REVIEW

File: _posts/2016-12-19-introduction-to-progressive-web-apps-push-notifications-part-3.markdown [hunk 1]
```diff
@@ -0,0 +1,971 @@
...
+//Push notification button
+  var fabPushElement = document.querySelector('.fab__push');
+  var fabPushImgElement = document.querySelector('.fab__image');
+```
+
+![Push Notification button](https://cdn.auth0.com/blog/pwa/notification-button.png)
+
+_Push Notification button_
+
+This code above simply grabs the push notification activation and deactivation button.
```
Comment (kukicado) [inline _posts/2016-12-19-introduction-to-progressive-web-apps-push-notifications-part-3.markdown]: Delete "above"

File: _posts/2016-12-19-introduction-to-progressive-web-apps-push-notifications-part-3.markdown [hunk 2]
```diff
@@ -0,0 +1,971 @@
...
+        })
+        .catch(function (error) {
+          console.error('Error occurred while enabling push ', error);
+        });
+      });
+  }
+
+```
+
+This code above checks the browser to determine whether push notification is supported. Now, it's paramount that the service worker has to be registered and ready before you can even try to subscribe a user to receive push notifications. So, the code above also checks if the service worker is ready and gets the subscription of the user.
```
Comment (kukicado) [inline _posts/2016-12-19-introduction-to-progressive-web-apps-push-notifications-part-3.markdown]: Delete "above"

File: _posts/2016-12-19-introduction-to-progressive-web-apps-push-notifications-part-3.markdown [hunk 3]
```diff
@@ -0,0 +1,971 @@
...
+      })
+      .catch(function (error) {
+        console.error('Failed to unsubscribe push notification.');
+      });
+    })
+  }
+
+```
+
+This code above is responsible for unsubscribing from push notification. A toast message indicates the unsubscription, then goes ahead to change the color of the button and delete the subscription ID.
```
Comment (kukicado) [inline _posts/2016-12-19-introduction-to-progressive-web-apps-push-notifications-part-3.markdown]: delete "above"

File: _posts/2016-12-19-introduction-to-progressive-web-apps-push-notifications-part-3.markdown [hunk 4]
```diff
@@ -0,0 +1,971 @@
...
+      unsubscribePush();
+    }
+    else {
+      subscribePush();
+    }
+  });
+
+```
+
+This code above simply adds a click event to the button to toggle between subscribing and unsubscribing a user.
```
Comment (kukicado) [inline _posts/2016-12-19-introduction-to-progressive-web-apps-push-notifications-part-3.markdown]: Delete "above"

File: _posts/2016-12-19-introduction-to-progressive-web-apps-push-notifications-part-3.markdown [hunk 5]
```diff
@@ -0,0 +1,971 @@
...
+
+### API Service
+
+The API service handling the saving and deletion of subscription IDs will also handle the actual sending of notifications. This is a break down of the API. It will have 3 api routes/endpoints:
+
+* `POST /api/users` to create new users and store their subscription IDs
+* `DELETE /api/user/:user_id` to delete and unsubscribe users
+* `POST /api/notify` to send notifications to all subscribed users
+
+Lucky enough, I have coded the [API service](https://github.com/unicodeveloper/pwa-api). Make sure you have `node` and `mongodb` installed. Clone it and run from the terminal with `node server.js` like so:
```
Comment (kukicado) [inline _posts/2016-12-19-introduction-to-progressive-web-apps-push-notifications-part-3.markdown]: Does the user need to setup mongo in any specific way? If so, i'd add the explanation here.

File: _posts/2016-12-19-introduction-to-progressive-web-apps-push-notifications-part-3.markdown [hunk 6]
```diff
@@ -0,0 +1,971 @@
...
+```
+
+Here, the code above listens to the event that is fired when a user clicks on the notification. `event.notification.close()`, closes the notification once clicked. Then, a new window or tab will be opened re-directing to `localhost:8080/latest.html.`
+
+**Note:** `event.waitUntil()` is been called to ensure the browser doesn't terminate our service worker before our new window has been displayed.
+
+
+### Automate Notification Sending Process
+
+We have been manually making a post request via *postman*. Practically, we want the user to get a notification once a commit has been made to the Github repository, `https://github.com/unicodeveloper/resources-i-like/`. So, how do we automate this process?
```
Comment (kukicado) [inline _posts/2016-12-19-introduction-to-progressive-web-apps-push-notifications-part-3.markdown]: Capitalize P in postman.

File: _posts/2016-12-19-introduction-to-progressive-web-apps-push-notifications-part-3.markdown [hunk 7]
```diff
@@ -0,0 +1,971 @@
...
+
+![Auth0 Lock Screen](https://cdn.auth0.com/blog/nexthrone-auth0lock.png)
+
+_Auth0 Lock Screen_
+
+In the case of an offline-first app, authenticating the user against a remote database won't be possible when network connectivity is lost. However, with service workers, you have full control over which pages and scripts are loaded when the user is offline. This means you can configure your `offline.html` file to display a useful message stating the user needs to regain connectivity to login again instead of displaying the Lock login screen.
+
+## Conclusion
+
+We have been able to successfully make our app work offline, load instantly, receive push notifications and also installable.
```
Comment (kukicado) [inline _posts/2016-12-19-introduction-to-progressive-web-apps-push-notifications-part-3.markdown]: and comma between notifications and and

File: _posts/2016-12-19-introduction-to-progressive-web-apps-push-notifications-part-3.markdown [hunk 8]
```diff
@@ -0,0 +1,971 @@
...
+
+I recommend that you use this tool frequently when developing a progressive web app. This tutorial wouldn't have been possible without [Ire's series on PWA](https://bitsofco.de/bitsofcode-pwa-part-1-offline-first-with-service-worker), [Timi's server side push notification tutorial](https://chunksofco.de/push-notifications-on-the-web-building-a-pwa-crossover-20f0317987de#.x311m3x56), [Gokulakrishnan's PWA demo app](https://github.com/gokulkrishh) and the guys at Google that work and blog daily about progressive web apps. Thanks a bunch!
+
+Hopefully you're now ready to dive fully into make your web applications progressive!
```
Comment (kukicado) [inline _posts/2016-12-19-introduction-to-progressive-web-apps-push-notifications-part-3.markdown]: I would add a link to the code repo at the end that would have all 3 parts and show the completed application. Great work on this! Very insightful!

File: _posts/2016-12-19-introduction-to-progressive-web-apps-push-notifications-part-3.markdown [hunk 9]
```diff
@@ -0,0 +1,971 @@
...
+
+---
+
+**TL;DR:** Web development has evolved significantly over the years allowing developers to deploy a website or web application and serve millions of people around the globe within minutes. With just a browser, a user can put in a URL and access a web application. With, **Progressive Web Apps**, developers can deliver amazing app-like experiences to users using modern web technologies. In [Part 1](https://auth0.com/blog/introduction-to-progressive-apps-part-one) and [Part 2](part 2) of this tutorial, we set up our progressive web app, cached the pages and made it work fully offline. This time, we'll add the ability to activate push notifications.
+
+---
+
+## Recap and Introduction to Part 3
+
+In [Introduction to Progressive Web Apps (Offline First)](https://auth0.com/blog/introduction-to-progressive-apps-part-one), we discussed how a typical progressive web application should look like and also introduced the service worker. We also cached the application shell. In [Introduction to Progressive Web Apps (Instant Loading)](https://link-to-part-2), we made the app cache dynamic data and load instantly from locally saved data.
```
Comment (unicodeveloper) [inline _posts/2016-12-19-introduction-to-progressive-web-apps-push-notifications-part-3.markdown]: Will have to wait till part 2 is published

File: _posts/2016-12-19-introduction-to-progressive-web-apps-push-notifications-part-3.markdown (hunk 8 above)
Comment (unicodeveloper) [inline _posts/2016-12-19-introduction-to-progressive-web-apps-push-notifications-part-3.markdown]: Already added the code repo in the earlier section `Host PWA`

**Pull request merged**